### PR TITLE
[NNVM] Add argmax and argmin operations from topi

### PR DESCRIPTION
**Pull request by grwlf**
Makes argmax operation accessible from NNVM.

There is an issue regarding output types, since the test seems to treats integer value, returned by argmax, as a floating-point. I'd like to ask for advice here. Probably, graph_runtime requires additional changes.

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for the contribution, please request reviews from reviewers

File: nnvm/tests/python/compiler/test_top_level4.py [hunk 1]
```diff
@@ -114,6 +116,22 @@ def test_reduce():
     verify_reduce((4, 4, 3), np.sum, sym.sum, axis=(0, 2))
     verify_reduce((4, 4, 3), np.sum, sym.sum)
 
+    def _with_keepdims(f):
```
Comment (nishi-t) [inline nnvm/tests/python/compiler/test_top_level4.py]: Just a matter of taste, but I'd suggest that the _with_keepdims definition is before all verify_reduce

```
def test_reduce():
    def _with_keepdims(f):
        ...
    verify_reduce((2, 3, 4), np.max, sym.max, axis=1, keepdims=True)
    ...
```

File: nnvm/tests/python/compiler/test_top_level4.py [hunk 2]
```diff
@@ -114,6 +116,22 @@ def test_reduce():
...
+                shape = list(a.shape)
+                shape.insert(axis,1)
+                a = a.reshape(tuple(shape))
+            return a
+        return wrapper
+
+    verify_reduce((4, 4, 3), _with_keepdims(np.argmax), sym.argmax, otype='int32')
+    verify_reduce((4, 4, 3), _with_keepdims(np.argmax), sym.argmax, oshape=(4,1,3), otype='int32', axis=1, keepdims=True)
+    verify_reduce((4, 4, 3), _with_keepdims(np.argmin), sym.argmin, otype='int32')
+    verify_reduce((4, 4, 3), _with_keepdims(np.argmin), sym.argmin, oshape=(4,1,3), otype='int32', axis=1, keepdims=True)
```
Comment (nishi-t) [inline nnvm/tests/python/compiler/test_top_level4.py]: Could you add a pattern which  keepdims is True and axis is None?

File: nnvm/tests/python/compiler/test_top_level4.py (hunk 1 above)
Comment (nishi-t) [inline nnvm/tests/python/compiler/test_top_level4.py]: Just a matter of taste, but I'd suggest that the _with_keepdims definition is before all verify_reduce

```
def test_reduce():
    def _with_keepdims(f):
        ...
    verify_reduce((2, 3, 4), np.max, sym.max, axis=1, keepdims=True)
    ...
```

File: nnvm/src/top/tensor/reduce.cc [hunk 3]
```diff
@@ -262,5 +262,62 @@ NNVM_REGISTER_BASE_REDUCE_OP(collapse_sum)
...
+                          std::vector<int>* out_attrs) {
+  // Static type inference for argmax operation. Argmax return indices which
+  // should have Int32 type as shapes do.
+  CHECK_EQ(in_attrs->size(), 1U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  NNVM_ASSIGN_OUTPUT_TYPE(attrs, *out_attrs, 0, static_cast<int>(Type));
+  return true;
+}
+
+NNVM_REGISTER_BASE_REDUCE_OP(argmax)
```
Comment (nishi-t) [inline nnvm/src/top/tensor/reduce.cc]: `NNVM_REGISTER_BASE_REDUCE_OP` uses ReduceParam, but does argmax and argmin support the [exclude option](https://github.com/dmlc/tvm/blob/master/nnvm/include/nnvm/top/tensor.h#L227)?

File: nnvm/tests/python/compiler/test_top_level4.py [hunk 4]
```diff
@@ -114,6 +116,22 @@ def test_reduce():
     verify_reduce((4, 4, 3), np.sum, sym.sum, axis=(0, 2))
     verify_reduce((4, 4, 3), np.sum, sym.sum)
 
+    def _with_keepdims(f):
+        """ Wrapper around numpy's argmax/argmin with `keepdims` argument supported """
+        def wrapper(x, axis=None, keepdims=False):
+            a = f(x, axis=axis)
+            if keepdims:
+                shape = list(a.shape)
+                shape.insert(axis,1)
```
[the remark this answers is not in the document]
Comment (grwlf) [reply nnvm/tests/python/compiler/test_top_level4.py]: Thanks, I am about to fix it, but see my comment below.

File: nnvm/tests/python/compiler/test_top_level4.py (hunk 2 above)
Comment (grwlf) [reply nnvm/tests/python/compiler/test_top_level4.py]: Indeed, `axis=None` doesn't work, but the wrong shape is not the only reason. Looking into sources I would say that this value of axis is forbidden for every reduce operation in NNVM. Axis is defined as `struct ReduceParams { .. TShape axis; .. }`, but `FieldEntryBase::Set` doesn't have specializations for TShape and the default version rejects Nones with 

```
NNVMError: Invalid Parameter format for axis expect  but value='None'
```
(note also missing _type name)

I would prefer to not include None usecases in this PR, but solve this issue separately.

File: nnvm/tests/python/compiler/test_top_level4.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nishi-t) [reply nnvm/tests/python/compiler/test_top_level4.py]: Ok, never mind.

File: nnvm/tests/python/compiler/test_top_level4.py (hunk 2 above)
Comment (nishi-t) [reply nnvm/tests/python/compiler/test_top_level4.py]: Ok, I got it.

File: nnvm/src/top/tensor/reduce.cc (hunk 3 above)
Comment (grwlf) [reply nnvm/src/top/tensor/reduce.cc]: Done, please check.

Comment (tqchen) [reply nnvm/src/top/tensor/reduce.cc]: For now, it is nice to keep all reduction operator consistent, as I assume the logics are going to be the same

Comment (nishi-t) [reply nnvm/src/top/tensor/reduce.cc]: @tqchen Thank you for your comments. I got it.
@grwlf Thank you for your works. Please don't mind my above comment.

**Pull request merged**